kotlinx-coroutines-test/src/DelayController.kt

```diff
@@ -112,9 +113,75 @@ public interface DelayController {
      * Resumed dispatchers will automatically progress through all coroutines scheduled at the current time. To advance
      * time and execute coroutines scheduled in the future use, one of [advanceTimeBy],
      * or [advanceUntilIdle].
+     *
+     * When the dispatcher is resumed, all execution be immediate in the thread that triggered it. This means
+     * that the following code will not switch back from Dispatchers.IO after `withContext`
+     *
+     * ```
+     * runBlockingTest {
+     *     withContext(Dispatchers.IO) { doIo() }
+     *     // runBlockingTest is still on Dispatchers.IO here
+     * }
+     * ```
+     *
+     * For tests that need accurate threading behavior, [pauseDispatcher] will ensure that the following test dispatches
+     * on the correct thread.
+     *
+     * ```
+     * runBlockingTest {
+     *     pauseDispatcher()
+     *     withContext(Dispatchers.IO) { doIo() }
+     *     // runBlockingTest has returned to it's starting thread here
+     * }
+     * ```
      */
     @ExperimentalCoroutinesApi // Since 1.2.1, tentatively till 1.3.0
     public fun resumeDispatcher()
+
+    /**
+     * Represents the queue state of a DelayController.
+     *
+     * Tests do not normally need to use this API. It is exposed for advanced situations like integrating multiple
+     * [TestCoroutineDispatcher] instances or creating alternatives to [runBlockingtest].
+     */
+    public sealed class QueueState {
```

Comment:
It is for #1202, right?

If I was implementing this in some closed-world application, then I would write more or less the same solution, trying to abstract everything with interfaces and be as generic as possible.

But for the library, it is completely different. It should have as small as possible API surface, at the same time being elegant, approachable by novices and pragmatic (a lot of buzzwords here, sorry for that :))

I would rather remove this API and re-design it from scratch when we have a couple of real world problems that cannot be elegantly solved with existing API for tests

Reply:
Agreed  - I put it up for review before writing tests because I wasn't sure this was the right direction to take the API.

This logic can by making TestCoroutineDispatcher implement EventLoop which is probably the preferred approach.

This code will be backed out.